File: client/lib/models/apartment_model.dart
class Apartment {
  final String? id;
  final String apartmentNumber;
  final String address;
  final String? overview;
  final double rentalPrice;
  final int floor;
  final int rooms;
  final int maxCapacity;
  final bool liftAvailable;
  final String contactNumber;
  final List<double> location;
  final String owner;
  final List<String> images;

  Apartment({
    this.id,
    required this.apartmentNumber,
    required this.address,
    this.overview,
    required this.rentalPrice,
    required this.floor,
    required this.rooms,
    required this.maxCapacity,
    required this.liftAvailable,
    required this.contactNumber,
    required this.location,
    required this.owner,
    required this.images,
  });

  factory Apartment.fromJson(Map<String, dynamic> json) {
    return Apartment(
      id: json['_id'],
      apartmentNumber: json['apartment_number'],
      address: json['address'],
      overview: json['overview'],
      rentalPrice: json['rental_price'].toDouble(),
      floor: json['floor'],
      rooms: json['rooms'],
      maxCapacity: json['max_capacity'],
      liftAvailable: json['liftAvailable'],
      contactNumber: json['contact_number'],
      location: List<double>.from(json['location']['coordinates']),
      owner: json['owner'],
      images: List<String>.from(json['images']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'apartment_number': apartmentNumber,
      'address': address,
      'overview': overview,
      'rental_price': rentalPrice,
      'floor': floor,
      'rooms': rooms,
      'max_capacity': maxCapacity,
      'liftAvailable': liftAvailable,
      'contact_number': contactNumber,
      'location': {'coordinates': location},
      'owner': owner,
      'images': images,
    };
  }
}

File: server/models/apartmentModel.js
const mongoose = require('mongoose');

const apartmentSchema = new mongoose.Schema({
    apartment_number: {
        type: String,
        required: true
    },
    address: {
        type: String,
        required: true
    },
    overview: {
        type: String
    },
    rental_price: {
        type: Number,
        required: true
    },
    floor: {
        type: Number,
        required: true
    },
    rooms: {
        type: Number,
        required: true
    },
    max_capacity: {
        type: Number,
        required: true
    },
    liftAvailable: {
        type: Boolean,
        default: false
    },
    contact_number: {
        type: String,
        required: true
    },
    location: {
        type: {
            type: String,
            default: 'Point',
        },
        coordinates: {
            type: [Number],
            required: true
        }
    },
    owner: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    images: {
        type: [String],
    }
});

apartmentSchema.index({ location: '2dsphere' });

module.exports = mongoose.model('Apartment', apartmentSchema);